Add scripts to export function metrics (JSON/CSV) and function call graph (DOT)

itxsamad1 opened the pull request:
## Summary

- Adds three self-contained Base scripts for exporting analysis results:
  - `ExportFunctionMetricsToJson`: per-function metrics to JSON with a summary section and optional selection/highlight filtering.
  - `ExportFunctionMetricsToCsv`: same metrics to CSV with a stable header.
  - `ExportFunctionCallGraphToDot`: exports the program function call graph to Graphviz DOT; supports selection/highlight filtering.

## Motivation

- Provide lightweight, automation-friendly exports for CI, baselining, and external tooling without changing core UI or build.
- Complements existing scripts (e.g., ComputeCyclomaticComplexity) with batch reporting and a simple call graph exporter.

## Details

- Metrics include:
  - identity: `name`, `entry`, `namespace`
  - body stats: `size_bytes` (address count), `address_ranges`, `instruction_count`, `basic_blocks`
  - complexity/interop: `cyclomatic_complexity`, `parameters`, `locals`
  - flags: `external`, `thunk`, `variadic`, `inline`, `no_return`, `custom_storage`
  - calling: `stack_purge_size`, `calling_convention`, `callers`, `callees`
  - types/signature: `return_type`, `prototype`, `prototype_with_cc`
- Selection/highlight filtering limits output to functions intersecting the chosen scope.
- Progress monitors update and respect cancellation; computation errors are logged and do not abort export.
- DOT exporter labels nodes as `functionName\nentryAddress` and writes simple edges from callers to callees.

## Usage

- Run from Script Manager on an analyzed program:
  - ExportFunctionMetricsToJson → choose `.json`
  - ExportFunctionMetricsToCsv → choose `.csv`
  - ExportFunctionCallGraphToDot → choose `.dot`
- Optional: select or highlight an address range before running to filter output.

## Files Changed

- `Ghidra/Features/Base/ghidra_scripts/ExportFunctionMetricsToJson.java:1`
- `Ghidra/Features/Base/ghidra_scripts/ExportFunctionMetricsToCsv.java:1`
- `Ghidra/Features/Base/ghidra_scripts/ExportFunctionCallGraphToDot.java:1`

## How To Test

- Open any program with analysis applied.
- Optionally select a subset of code to test filtering.
- Run each script; verify:
  - JSON emits a valid document with `program`, `functions[]`, and `summary`.
  - CSV imports cleanly (quoted header and fields).
  - DOT renders via Graphviz, e.g., `dot -Tpng out.dot -o out.png`, and reflects expected edges.

## Scope / Compliance

- No 3rd-party dependencies; no build system changes.
- Small, focused scripts under Base `ghidra_scripts`.
- Adheres to CONTRIBUTING guidance for small, utility-focused improvements.

## Notes

- `size_bytes` reports address count in the function body; `address_ranges` clarifies segmentation.
- Locals are omitted for external functions.
- DOT exporter keeps edges within selection when filtering is used.

## Optional Checklist

- [ ] Ran scripts on at least one sample program
- [ ] Validated JSON structure and CSV header in common tools
- [ ] Rendered DOT via Graphviz
- [ ] Scoped PR to scripts only; no refactors or external jars

itxsamad1 committed:
Add scripts: export function metrics (JSON/CSV) and function call graph (DOT)

ryanmkurtz reviewed (changes requested):
I think it would make more sense to add this functionality to the existing [ExportFunctionInfoScript](https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/Base/ghidra_scripts/ExportFunctionInfoScript.java).  `GhidraScript.askChoices()` could be used to choose the export format.

itxsamad1 committed:
Integrate function export scripts into ExportFunctionInfoScript

- Consolidated ExportFunctionMetricsToJson, ExportFunctionMetricsToCsv, and ExportFunctionCallGraphToDot into ExportFunctionInfoScript.java
- Added format selection using GhidraScript.askChoices() as requested in code review
- Users can now choose between: JSON (Simple), JSON (Detailed metrics), CSV (Detailed metrics), and DOT (Function call graph)
- All export formats support optional selection/highlight filtering
- Maintains backward compatibility with existing simple JSON export functionality

ryanmkurtz commented:
Thanks, I hope to review it next week.

itxsamad1 committed:
Address code review feedback: fix lambda variable issue, remove unused variable, propagate CancelledException, use for-each loops

ryanmkurtz commented on Ghidra/Features/Base/ghidra_scripts/ExportFunctionInfoScript.java:
```diff
@@ -192,10 +175,9 @@ private void exportMetricsJson(File outputFile, AddressSetView scope) throws IOE
 				int instCount = 0;
-				for (Iterator<Instruction> insIt = listing.getInstructions(body, true); insIt.hasNext();) {
-					insIt.next();
+				for (Instruction inst : listing.getInstructions(body, true)) {
```
This introduces a new warning because `inst` is unused.  It was probably better the way it was before.  Even better, you can try using a stream...something like `long instCount = CollectionUtils.asStream(listing.getInstructions(body, true)).count()`.  I haven't tested that though...you'd need to verify.

ryanmkurtz commented on Ghidra/Features/Base/ghidra_scripts/ExportFunctionInfoScript.java:
What exception is this ignoring?

itxsamad1 commented:
Superseded by #9311, which extends ExportFunctionInfoScript with the consolidated export formats. All review feedback from this PR has been addressed in that follow-up (Gson-based JSON metrics, monitor.checkCancelled() / CancelledException propagation, DOT graph labeling, and related cleanup).

https://github.com/NationalSecurityAgency/ghidra/pull/9311
